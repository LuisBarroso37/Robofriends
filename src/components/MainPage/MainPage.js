import React, { Component } from 'react';
import CardList from '../Cardlist/Cardlist';
import SearchBox from '../SearchBox/SearchBox';
import Scroll from '../Scroll/Scroll';
import ErrorBoundary from '../ErrorBoundary/ErrorBoundary';
import Header from '../Header/Header';
import './MainPage.css';

class MainPage extends Component {
    componentDidMount() {
        this.props.onRequestRobots();
    }

    filterRobots = () => {
        const { robots, searchField } = this.props;
        console.log(robots);
        return robots.filter(robot => {
            return robot.name.toLowerCase().includes(searchField.toLowerCase());
        });
    }

    render() {
        const { onSearchChange, isPending } = this.props

        return (
            <div className='tc'>
                <Header />
                <SearchBox searchChange={onSearchChange}/>
                <Scroll>
                    { isPending ? <h1>Loading...</h1> :
                        <ErrorBoundary>
                            <CardList robots={this.filterRobots()} />
                        </ErrorBoundary>
                    }
                </Scroll>
            </div>
        );
        }
    }

export default MainPage;