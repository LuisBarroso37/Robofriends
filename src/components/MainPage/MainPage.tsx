import React, { Component } from 'react';
import CardList from '../Cardlist/Cardlist';
import SearchBox from '../SearchBox/SearchBox';
import Scroll from '../Scroll/Scroll';
import ErrorBoundary from '../ErrorBoundary/ErrorBoundary';
import Header from '../Header/Header';
import './MainPage.css';
import { Props, IRobot } from '../../interfaces';

class MainPage extends Component<Props, {}> {
    componentDidMount(): void {
        this.props.onRequestRobots();
    }

    filterRobots = () : Array<IRobot> => {
        const { robots, searchField } = this.props;
        
        return robots.filter(robot => {
            return robot.name.toLowerCase().includes(searchField.toLowerCase());
        });
    }

    render(): JSX.Element {
        const { onSearchChange, isPending } = this.props

        return (
            <div className='tc'>
                <Header />
                <SearchBox searchChange={onSearchChange}/>
                <Scroll>
                    { isPending ? <h1>Loading...</h1> :
                        <ErrorBoundary>
                            <CardList robots={this.filterRobots()} />
                        </ErrorBoundary>
                    }
                </Scroll>
            </div>
        );
        }
    }

export default MainPage;