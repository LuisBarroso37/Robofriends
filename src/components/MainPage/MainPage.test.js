import { shallow } from 'enzyme';
import MainPage from './MainPage';
import React from 'react';

let wrapper;

describe('MainPage component', () => {
    // Ran before each test
    beforeEach(() => {
        const mockProps = {
            onRequestRobots: jest.fn(),
            robots: [],
            searchField: '',
            isPending: false
        }

        wrapper = shallow(<MainPage {...mockProps}/>);
    });

    it('render MainPage without crashing', () => {
        expect(wrapper).toMatchSnapshot();
    });

    it('Filter robots correctly', () => {
        const mockProps2 = {
            onRequestRobots: jest.fn(),
            robots: [{
                id: 3,
                name: 'John',
                email: 'john@gmail.com'
            }],
            searchField: 'john',
            isPending: false
        }
        const wrapper2 = shallow(<MainPage {...mockProps2}/>);
        expect(wrapper2.instance().filterRobots()).toEqual(mockProps2.robots);
    });

    it('Filter robots null when searchField doesn\'t match any robot', () => {
        const mockProps2 = {
            onRequestRobots: jest.fn(),
            robots: [{
                id: 3,
                name: 'John',
                email: 'john@gmail.com'
            }],
            searchField: 'a',
            isPending: false
        }
        const wrapper2 = shallow(<MainPage {...mockProps2}/>);
        expect(wrapper2.instance().filterRobots()).toEqual([]);
    });

    it('different render when isPending is true', () => {
        const mockProps3 = {
            onRequestRobots: jest.fn(),
            robots: [{
                id: 3,
                name: 'John',
                email: 'john@gmail.com'
            }],
            searchField: 'a',
            isPending: true
        };

        wrapper = shallow(<MainPage {...mockProps3}/>);

        expect(wrapper).toMatchSnapshot();
    });
});