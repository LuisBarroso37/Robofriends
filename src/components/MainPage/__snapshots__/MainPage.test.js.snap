// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainPage component different render when isPending is true 1`] = `ShallowWrapper {}`;

exports[`MainPage component render MainPage without crashing 1`] = `ShallowWrapper {}`;
